src/ts/Document.ts

```diff
@@ -0,0 +1,223 @@
+import { assignIn, cloneDeep, get as getIn, set as setIn } from 'lodash';
+import { ForeignKeyDbTransformer } from '#shared/classes';
+import { TransformDataError } from '#shared/errors';
+import type { GenericObject, MvRecord } from '#shared/types';
+import type Schema from './Schema';
+
+const DEFAULT_PROPERTY_DESCRIPTOR: PropertyDescriptor = {
+	configurable: false,
+	enumerable: false,
+	writable: false,
+};
+
+export interface DocumentConstructorOptionsData {
+	data: GenericObject;
+	isSubdocument?: boolean;
+}
+export interface DocumentConstructorOptionsRecord {
+	record: MvRecord;
+	isSubdocument?: boolean;
+}
+export interface DocumentConstructorOptionsRaw {
+	record: MvRecord;
+}
+export type DocumentConstructorOptions =
+	| DocumentConstructorOptionsData
+	| DocumentConstructorOptionsRecord
+	| DocumentConstructorOptionsRaw;
+
+export interface BuildForeignKeyDefinitionsResult {
+	filename: string;
+	entityName: string;
+	entityIds: string[];
+}
+
+/** A document object */
+class Document {
+	[key: string]: unknown;
+
+	public _id?: string;
+
+	public _raw?: MvRecord;
+
+	/** Array of any errors which occurred during transformation from the database */
+	public transformationErrors: TransformDataError[];
+
+	/** Schema instance which defined this document */
+	private readonly schema: Schema | null;
+
+	/** Record array of multivalue data */
+	private record: MvRecord;
+
+	/** Indicates whether this document is a subdocument of a composing parent */
+	private readonly isSubdocument: boolean;
+
+	public constructor(schema: Schema, options: DocumentConstructorOptions);
```

Comment:
I've added overloads here to try to ensure that the constructor is called appropriately.